### DISCUSSION TITLE
Integrating Boost

### DISCUSSION THREAD

**Discussion by alepping** [nebulastream/nebulastream#142] [Ideas]
# Description
NebulaStream used boost in the past, at least for serialization, but removed it because it was [cumbersome](https://github.com/nebulastream/nebulastream/issues/667) in contrast to "totally rely on protobuf".
Ultimately, Boost was purged because [it was not used anywhere](https://github.com/nebulastream/nebulastream/issues/964).

## Upsides
We might want to reconsider using boost for the following reasons (the below list is a wip):
- reading from TCP/UDP sockets
- writing to TCP/UDP sockets
- testing tcp/udp sources and sinks
- using boost::asio to spawn a multi-threaded file-socket server that pushes data into sources and potentially improves on the single-thread per source design
- boost provides a lot of string processing functionality, including a JSON parser and more or less a CSV parser|

## Downsides
- we need to ship the boost dependencies with our project (we can use vcpkg)
- potentially a single boost dependency (e.g., boost::asio) is very large (todo: find out how large)

**Comment by PhilippGrulich** [comment]
Just my two cents from the past. We mainly removed boost because it was a large pice of shit. Similar to folly. I think these libs sound nice because they offer many nice abstractions and utilities but this comes with a certain binary and maintenance price;) 
The problem is, that we first introduce them to a very specific use-case. There everything is nice and isolated. But than you wake up and see that now boost is used everywhere and you have a hard time to remove it. My learning from this is to be very carful with third party dependency, use as small and few as possible and with a clear single use-case. 

Sorry for the rant ;)

**Reply by alepping** [reply]
@PhilippGrulich 
Was it possible for you to include just parts of boost or did you include the entire boost project?
I just tested it and added just boost asio, which is a rather small project (few MBs) that minimally increases the binary size.

**Reply by PhilippGrulich** [reply]
Maybe if you limit it to asio it's fine 🤷. I just would be cautious and really check if it's the best solution. Also make sure everything is static linked and that users don't need any boost stuff. But vcpkg should ensure this.

**Reply by alepping** [reply]
Thanks for the rant btw! I really wanted to hear something like that :D

**Reply by PhilippGrulich** [reply]
> Thanks for the rant btw! I really wanted to hear something like that :D

Haha I think the word boost triggers some bad memories :D

**Reply by zeuchste** [reply]
Ok so how do we then proceed now?
